src/main/kotlin/brokers/PgBroker.kt

```diff
@@ -0,0 +1,153 @@
+    val scope: CoroutineScope = GlobalScope,
+
+    val emptyMessageDelay: Duration = 5000.milliseconds,
+    val messageReservationTimeoutMs: Duration = 5.minutes,
+): IMessageBroker {
+    companion object {
+        const val HEADERS_PREFIX = "kot-pg"
+        const val ID_HEADER = "${HEADERS_PREFIX}-id"
+    }
+
+    init {
+        transaction(dbWrapper.connection) {
+            SchemaUtils.createMissingTablesAndColumns(KotaskMessages)
+        }
+    }
+
+    override fun submitMessage(queueName: QueueName, message: Message): Unit {
+
+        transaction(dbWrapper.connection) {
+            val id = UUID.randomUUID()
+            KotaskMessages.insert {
+                it[this.id] = id
+                it[this.callId] = message.getCallId()
+                it[this.queueName] = queueName
+                it[this.body] = message.body
+                it[this.attemptNum] = message.getAttemptNum()
+                it[this.delayMs] = message.delayMs
+                it[this.scheduledAt] = Instant.now().plusMillis(message.delayMs)
+                it[this.headers] = Json.encodeToString(
+                    buildMap {
+                        putAll(message.headers)
+                        put(ID_HEADER, id.toString())
+                    }
+                ).encodeToByteArray()
+            }
+        }
+    }
+
+    override fun startConsumer(queueName: QueueName, handler: ConsumerHandler): IConsumer {
+        var isStopped = false
+        val job = scope.launch(MDCContext()) {
+            while (!isStopped) {
+                val messages = transaction(dbWrapper.connection) {
+                    val messages = KotaskMessages
+                        .selectAll()
+                        .andWhere {
+                            (KotaskMessages.queueName eq queueName) and
+                            (KotaskMessages.scheduledAt less Instant.now()) and
+                            (
+                                KotaskMessages.startedAt less Instant.now().minusSeconds(messageReservationTimeoutMs.inWholeSeconds) or // settings
+                                KotaskMessages.startedAt.isNull()
+                            )
+                        }
+                        .limit(100) // settings
+                        .forUpdate()
+                        .map {
+                            val headers = Json.decodeFromString<HashMap<String, String>>(it[KotaskMessages.headers].decodeToString())
+                            Message(
+                                body = it[KotaskMessages.body],
+                                headers = headers,
+                                delayMs = it[KotaskMessages.delayMs],
+                            )
+                        }
+
+                    val callIds = messages.map { it.getCallId() }
+
+                    KotaskMessages.update({
+                        KotaskMessages.callId inList callIds
+                    }) {
+                        it[startedAt] = Instant.now()
```

Comment:
It would be safe to always use the PG server time if it's possible. With the current implementation, if two consumers have different time, it can lead to a problem. What do you think?